run_cpu_exp3.py

```diff
@@ -0,0 +1,185 @@
+
+
+def fixed_generator_lenet(data_shape, label_shape):
+    data = np.random.rand(*data_shape)
+    labels = np.random.rand(*label_shape)
+
+    def yielder():
+        while True:
+            yield (data, labels)
+
+    return yielder
+
+
+def train(train_data_generator, val_data_generator, epochs, num_threads=None) -> int:
+    """
+    Trains model over training / validation data generators.
+    We measure the average time taken to train & validate the model for each epoch
+    """
+    if num_threads:
+        import tensorflow as tf
+
+        tf.config.threading.set_intra_op_parallelism_threads(num_threads)
+        tf.config.threading.set_inter_op_parallelism_threads(num_threads)
+
+    import keras
+
+    # class TimeHistory(keras.callbacks.Callback):
+    #     def on_train_begin(self, logs={}):
+    #         self.times = []
+
+    #     def on_epoch_begin(self, batch, logs={}):
+    #         self.epoch_time_start = time.time()
+
+    #     def on_epoch_end(self, batch, logs={}):
+    #         self.times.append(time.time() - self.epoch_time_start)
+
+    model = Lenet5().get_model()
+    opt = keras.optimizers.Adadelta()
+    model.compile(optimizer=opt, loss="mean_squared_error", metrics=["accuracy"])
+
+    time_callback = TimedCallback()
+
+    # time_callback = TimeHistory()
+    model.fit_generator(
+        generator=train_data_generator(),
+        steps_per_epoch=1,
+        epochs=epochs,
+        validation_data=val_data_generator(),
+        validation_steps=1,
+        callbacks=[time_callback]
+        # max_queue_size=3,
+        # workers=3,
+        # use_multiprocessing=True,
+    )
+    return time_callback.get_avg_epoch_running_time()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Experiment 3 - Max throughput on scaling machine unit"
+    )
+    parser.add_argument(
+        "training_rows",
+        metavar="tr",
+        type=int,
+        nargs="?",
+        help="Batch size on each generation",
+    )
+    parser.add_argument(
+        "sla", metavar="sla", type=float, nargs="?", help="SLA timing in seconds"
+    )
+    # parser.add_argument(
+    #     "executors",
+    #     metavar="e",
+    #     type=int,
+    #     nargs="?",
+    #     help="Number of executors used for distributed Spark training",
+    # )
+    parser.add_argument(
+        "epochs", metavar="ep", type=int, nargs="?", help="Training epochs"
+    )
+    parser.add_argument(
+        "validation_ratio",
+        metavar="vr",
+        type=float,
+        nargs="?",
+        help="Ratio of training data used for validation",
+    )
+    parser.add_argument(
+        "num_threads",
+        metavar="nt",
+        type=int,
+        nargs="?",
+        help="Number of threads used for training",
+    )
+    args = parser.parse_args()
+
+    TRAINING_ROWS = args.training_rows
+    VALIDATION_RATIO = args.validation_ratio
+    VALIDATION_ROWS = int(VALIDATION_RATIO * TRAINING_ROWS)
```

Comment:
We agreed that if our train : val ratio is 9 : 1, then if we have 10K training rows, we will train on 9K & validate on 1K. This doesn't seem like the implementation of the logic